File: qqbot-plugin-api/src/main/java/com/mieai/qqbot/plugin/api/CancellationToken.java
package com.mieai.qqbot.plugin.api;

import java.util.Objects;
import java.util.concurrent.CancellationException;
import java.util.concurrent.atomic.AtomicBoolean;

/** Cooperative cancellation signal for one plugin invocation. */
public final class CancellationToken {
    private static final ThreadLocal<CancellationToken> CURRENT = new ThreadLocal<>();
    private final AtomicBoolean cancelled;

    CancellationToken(AtomicBoolean cancelled) {
        this.cancelled = Objects.requireNonNull(cancelled, "cancelled must not be null");
    }

    /** Returns the invocation token on a host callback thread, or a live no-op token otherwise. */
    public static CancellationToken current() {
        CancellationToken token = CURRENT.get();
        return token == null ? NEVER : token;
    }

    public boolean isCancellationRequested() {
        return cancelled.get();
    }

    public void throwIfCancellationRequested() {
        if (isCancellationRequested()) {
            throw new CancellationException("Plugin invocation was cancelled");
        }
    }

    /** Host hook used to bind a token to a synchronous callback. */
    public static Scope activate(CancellationToken token) {
        CancellationToken previous = CURRENT.get();
        CURRENT.set(Objects.requireNonNull(token, "token must not be null"));
        return () -> {
            if (previous == null) CURRENT.remove(); else CURRENT.set(previous);
        };
    }

    public interface Scope extends AutoCloseable {
        @Override
        void close();
    }

    private static final CancellationToken NEVER = new CancellationToken(new AtomicBoolean());
}

File: qqbot-plugin-api/src/main/kotlin/com/mieai/qqbot/plugin/api/CancellationToken.kt
package com.mieai.qqbot.plugin.api

import java.util.concurrent.CancellationException
import java.util.concurrent.atomic.AtomicBoolean

/** Cooperative cancellation signal for one plugin invocation. */
class CancellationToken internal constructor(
    private val cancelled: AtomicBoolean,
) {
    /** Returns the invocation token on a host callback thread, or a live no-op token otherwise. */
    val isCancellationRequested: Boolean
        get() = cancelled.get()

    fun throwIfCancellationRequested() {
        if (isCancellationRequested) {
            throw CancellationException("Plugin invocation was cancelled")
        }
    }

    /** Host hook used to bind a token to a synchronous callback. */
    fun interface Scope : AutoCloseable {
        override fun close()
    }

    companion object {
        private val currentToken = ThreadLocal<CancellationToken>()
        private val never = CancellationToken(AtomicBoolean())

        fun current(): CancellationToken = currentToken.get() ?: never

        fun activate(token: CancellationToken): Scope {
            val previous = currentToken.get()
            currentToken.set(token)
            return Scope {
                if (previous == null) {
                    currentToken.remove()
                } else {
                    currentToken.set(previous)
                }
            }
        }
    }
}
